flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/StreamCommonSubGraphBasedOptimizer.scala

```diff
@@ -46,62 +47,98 @@ import scala.collection.JavaConversions._
 class StreamCommonSubGraphBasedOptimizer(planner: StreamPlanner)
   extends CommonSubGraphBasedOptimizer {
 
+  private def shouldSkipMiniBatch(blocks: Seq[RelNodeBlock]): Boolean = {
+
+    val noMiniBatchRequired = {
+      (node: RelNode) =>
+        node match {
+          case _: Filter | _: Project | _: TableScan | _: Calc | _: Values | _: Sink |
+              _: LegacySink =>
+            true
+          case unionNode: LogicalUnion => unionNode.all
+          case _ => false
+        }
+    }
+
+    def nodeTraverser(node: RelNode): Boolean = {
+      noMiniBatchRequired(node) && node.getInputs
+        .map(n => nodeTraverser(n))
+        .forall(r => r)
+    }
+
+    blocks.map(b => nodeTraverser(b.outputNode)).forall(r => r)
+  }
+
   override protected def doOptimize(roots: Seq[RelNode]): Seq[RelNodeBlock] = {
     val tableConfig = planner.getTableConfig
     // build RelNodeBlock plan
     val sinkBlocks = RelNodeBlockPlanBuilder.buildRelNodeBlockPlan(roots, tableConfig)
     // infer trait properties for sink block
-    sinkBlocks.foreach {
-      sinkBlock =>
-        // don't require update before by default
-        sinkBlock.setUpdateBeforeRequired(false)
-
-        val miniBatchInterval: MiniBatchInterval =
-          if (tableConfig.get(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ENABLED)) {
-            val miniBatchLatency =
-              tableConfig.get(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ALLOW_LATENCY).toMillis
-            Preconditions.checkArgument(
-              miniBatchLatency > 0,
-              "MiniBatch Latency must be greater than 0 ms.",
-              null)
-            new MiniBatchInterval(miniBatchLatency, MiniBatchMode.ProcTime)
-          } else {
-            MiniBatchIntervalTrait.NONE.getMiniBatchInterval
-          }
-        sinkBlock.setMiniBatchInterval(miniBatchInterval)
-    }
 
-    if (sinkBlocks.size == 1) {
-      // If there is only one sink block, the given relational expressions are a simple tree
-      // (only one root), not a dag. So many operations (e.g. infer and propagate
-      // requireUpdateBefore) can be omitted to save optimization time.
-      val block = sinkBlocks.head
-      val optimizedTree = optimizeTree(
-        block.getPlan,
-        block.isUpdateBeforeRequired,
-        block.getMiniBatchInterval,
-        isSinkBlock = true)
-      block.setOptimizedPlan(optimizedTree)
-      return sinkBlocks
-    }
+    val origMiniBatchEnabled = tableConfig.get(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ENABLED)
+
+    try {
```

Comment:
Introducing a huge try-finally block, in my opinion, isn't particularly good. How about introducing a function like `optimizeSinkBlocks` to encapsulate the statements inside the try block?